aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
   return at::any(self._values());
 }
 
+Tensor bmm_sparse_cpu(const SparseTensor& self, const Tensor& mat2) {
+  Tensor result = at::empty({}, mat2.options());
+  return bmm_out_sparse_cpu(result, self, mat2);
+}
+
+// Search a sorted strided array for the rightmost instance of a value.
+// Array must be sorted from lowest to highest
+template<typename scalar_t>
+scalar_t binary_search_strided_rightmost(scalar_t search_val, scalar_t* sorted_arr, int64_t sorted_arr_stride, int64_t length, bool* found) {
+  if (length == 0) {
+    *found = false;
+    return -1;
+  }
+
+  int64_t left_ind = 0;
+  int64_t right_ind = length - 1;
+  int64_t mid_ind;
+  bool done_searching = false;
+
+  while (!done_searching) {
+    mid_ind = (left_ind+right_ind) >> 1;
+    scalar_t mid_val = sorted_arr[mid_ind*sorted_arr_stride];
+
+    if (mid_val > search_val) {
+      right_ind = mid_ind-1;
+    } else if((mid_val == search_val) && (
+      (mid_ind == length - 1) || (sorted_arr[(mid_ind+1)*sorted_arr_stride] != search_val)
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
+
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  result.resize_({self.size(0), self.size(1), mat2.size(2)});
+
+  if (self._nnz() == 0) {
+    result.zero_();
+    return result;
+  }
+
+  // First need to coalesce to get all of the first dimension indices
+  // in order since we'll be sending each matrix into the MM operation
+  SparseTensor self_coalesced;
+
+  if (!self.is_coalesced()){
+    self_coalesced = self.coalesce();
```

Comment:
Huh, can't you just say `self_coalesced = self.coalesce()` unconditionally?

Reply:
The `coalesce()` call takes a pretty long time, even if the tensor is already coalesced. I don't remember how much time, but it was pretty dramatic in some cases.

Reply:
What do you mean by "long time"? 

```
SparseTensor coalesce_sparse_cpu(const SparseTensor& self) {
  AT_ASSERT(self.defined());
  TORCH_INTERNAL_ASSERT(at::impl::variable_excluded_from_dispatch());
  AT_ASSERT(self.is_sparse());
  
  if (self.is_coalesced()) {
    return self;
  }
```

and

```
SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
  int64_t nnz = self._nnz();
  if (self.is_coalesced()) {
    return self;
  }
```

So at most is should just be the cost of another dynamic dispatch.

Reply:
So if it is actually taking a long time we should figure out why.

Reply:
Ah you're right. I ran some direct measurements to compare whether my explicit `is_coalesced()` check saves any time. My claim that it's "pretty dramatic in some cases" was way off. There is a small speedup in most cases, but it's minute, usually around 1% or less. My guess is that this tiny speedup is coming from the fact that my explicit check avoids the three assertion checks in coalesce_sparse_cpu. I'll get rid of my explicit check since the gain is so small.


  |   |   |   |   | without checking |   | with checking |   | speedup with checking |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
device | num_matrices | squ_mat_elements | sparsity | workaround | not pre-coalesced | pre-coalesced | not pre-coalesced | pre-coalesced | not pre-coalesced | pre-coalesced
cpu | 10 | 100 | 0.9 | 0.001316 | 0.0006726 | 0.0004452 | 0.00067 | 0.0004418 | 1.003880597 | 1.00769579
cpu | 10 | 10000 | 0.999 | 0.001968 | 0.001083 | 0.0008478 | 0.0010672 | 0.0008358 | 1.014805097 | 1.014357502
cpu | 10 | 10000 | 0.99 | 0.002033 | 0.0014168 | 0.0009008 | 0.001391 | 0.0008846 | 1.018547807 | 1.018313362
cpu | 10 | 10000 | 0.9 | 0.002687 | 0.0048152 | 0.0013868 | 0.0048 | 0.0013716 | 1.003166667 | 1.011081948
cpu | 100 | 100 | 0.9 | 0.012417 | 0.004119 | 0.0036536 | 0.0041064 | 0.0036308 | 1.003068381 | 1.006279608
cpu | 100 | 10000 | 0.999 | 0.021109 | 0.0089946 | 0.0098808 | 0.0085026 | 0.0100254 | 1.057864653 | 0.9855766353
cpu | 100 | 10000 | 0.99 | 0.020283 | 0.0119564 | 0.0085836 | 0.0119212 | 0.0085624 | 1.002952723 | 1.002475941

Reply:
Great. This might mean we should maybe optimize `coalesce_sparse_Cpu` a little, but that can always be something done later.